MANUAL.adoc

```diff
@@ -3179,83 +3179,1155 @@ public class LoopExample {
+
+NOTE: While the Java language makes no distinction between a wildcard with no bound (`?`) and a wildcard with an `Object`-typed upper bound (`? extends Object`), the JVM specification does make such a distinction. Furthermore, the Java compiler `javac` will emit different attribute content in either case.
+
+[id=constructing-generic-types]
+=== Constructing generic types
+
+Instances of `GenericType`, are built
+using factory methods found on that class,
+or by creating derived instances using instance methods
+which yield new instances.
+Instances of `GenericType` are themselves immutable.
+
+Like `GenericType`, instances of `TypeArgument` are instantiated using
+factory methods or by acquiring derived instances from existing instances.
+
+//TEST:BEGIN
+.Example of constructing a complex generic type
+[source,java]
+----
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.List;
+import java.util.Map;
+
+import io.quarkus.gizmo2.GenericType;
+import io.quarkus.gizmo2.TypeArgument;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE_USE)
+@interface Silly {
+    int silliness() default 1;
+}
+
+@Retention(RetentionPolicy.CLASS)
+@Target(ElementType.TYPE_USE)
+@interface Serious {
+    int seriousness() default 1;
+}
+
+public class GenericTypeExample {
+    public static void main(String[] args) {
+        // simple type: @Serious int
+        GenericType.OfPrimitive seriousInt = GenericType.ofPrimitive(int.class).withAnnotation(Serious.class);
+        // array type: @Serious int @Silly []
+        GenericType.OfArray sillyArraySeriousInt = seriousInt.arrayType().withAnnotation(Silly.class);
+        // simple type: @Silly String
+        GenericType.OfClass sillyString = GenericType.ofClass(String.class).withAnnotation(Silly.class);
+        // type bound: @Serious(seriousness = 2) ? extends @Silly String
+        TypeArgument.OfExtends extendsSillyString =
+                TypeArgument.ofExtends(sillyString).withAnnotation(Serious.class, ac -> ac.add(Serious::seriousness, 2));
+        // type: @Silly List<@Serious(seriousness = 2) ? extends @Silly String
+        GenericType.OfClass sillyList =
+                GenericType.ofClass(List.class).withAnnotation(Silly.class).withArguments(List.of(extendsSillyString));
+        // construct the following complex type:
+        // Map<@Silly List<@Serious(seriousness = 2) ? extends @Silly String>, @Serious int @Silly []>
+        GenericType.OfClass mapType = GenericType.ofClass(Map.class)
+            .withArguments(List.of(
+                TypeArgument.ofExact(sillyList),
+                TypeArgument.ofExact(sillyArraySeriousInt)
+            ));
+        System.out.printf("Our map type is: %s%n", mapType);
+    }
+}
+----
+//TEST:RUN GenericTypeExample
```

Comment:
This should be just `//TEST:OUTPUT`. Although this probably also works, it didn't even occur to me...